library/src/main/java/com/datatorrent/lib/join/managed/ManagedTimeStateMultiMap.java

```diff
@@ -0,0 +1,247 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.lib.join.managed;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Future;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+
+import org.apache.apex.malhar.lib.state.managed.ManagedTimeStateImpl;
+import org.apache.apex.malhar.lib.state.spillable.Spillable;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+
+import com.datatorrent.api.StreamCodec;
+import com.datatorrent.lib.codec.KryoSerializableStreamCodec;
+import com.datatorrent.netlet.util.Slice;
+
+public class ManagedTimeStateMultiMap<K,V> implements Spillable.SpillableByteArrayListMultimap<K,V>
+{
+  private transient StreamCodec streamCodec = null;
+  private boolean isKeyContainsMultiValue = false;
+  private long timeBucket;
+  @NotNull
+  private ManagedTimeStateImpl store;
+
+  public ManagedTimeStateMultiMap()
+  {
+    if (streamCodec == null) {
+      streamCodec = new KryoSerializableStreamCodec();
+    }
+  }
+
+  public ManagedTimeStateMultiMap(@NotNull ManagedTimeStateImpl store, boolean isPrimaryKey)
+  {
+    this();
+    this.store = Preconditions.checkNotNull(store);
+    this.isKeyContainsMultiValue = isPrimaryKey;
+  }
+
+  @Override
+  public List<V> get(@Nullable K k)
+  {
+    List<V> value = null;
+    Slice valueSlice = store.getSync(getBucketId(k), streamCodec.toByteArray(k));
+    if (isKeyContainsMultiValue) {
+      value = (List<V>)streamCodec.fromByteArray(valueSlice);
+    }  else {
+      if (valueSlice == null || valueSlice.length == 0 || valueSlice.buffer == null) {
+        return null;
+      }
+      value = new ArrayList<>();
+      value.add((V)streamCodec.fromByteArray(valueSlice));
+    }
+    return  value;
+  }
+
+  public Future<Slice> getAsync(@Nullable K k)
```

Comment:
ideally it should return Future<List<V>>. can you write a composible future to implement it.